src/internet_identity/src/archive.rs

```diff
@@ -0,0 +1,179 @@
+            }
+        }
+    }
+}
+
+async fn create_and_install_archive(wasm: Vec<u8>) -> DeployArchiveResult {
+    match create_archive().await {
+        Err(e) => {
+            state::persistent_state_mut(|persistent_state| {
+                // unlock archive creation again
+                persistent_state.archive_info.state = ArchiveState::NotCreated
+            });
+            return CreationFailed(format!("failed to create archive: {:?}", e));
+        }
+        Ok(data) => {
+            let archive_canister_id = data.archive_canister;
+            state::persistent_state_mut(|persistent_state| {
+                // safe archive info permanently
+                persistent_state.archive_info.state = ArchiveState::Created(data)
+            });
+            match install_archive(archive_canister_id, wasm).await {
+                Ok(()) => DeployArchiveResult::Success(archive_canister_id),
+                Err(err) => UpgradeFailed(err),
+            }
+        }
+    }
+}
+
+async fn create_archive() -> CallResult<ArchiveData> {
+    let (result,) = create_canister(CreateCanisterArgument { settings: None }).await?;
+    Ok(ArchiveData {
+        sequence_number: 0,
+        archive_canister: result.canister_id,
+    })
+}
+
+async fn install_archive(archive_canister: Principal, wasm_module: Vec<u8>) -> Result<(), String> {
+    let expected_hash = verify_wasm_hash(&wasm_module)?;
+
+    let (archive_status,) = canister_status(CanisterIdRecord {
+        canister_id: archive_canister,
+    })
+    .await
+    .map_err(|err| format!("failed to retrieve archive status: {:?}", err))?;
+
+    let module_hash = archive_status.module_hash;
+    if module_hash
+        .clone()
+        .map(|hash| hash == expected_hash.to_vec())
+        .unwrap_or(false)
+    {
+        // Don't do anything further if the archive canister has the given module already installed
+        return Ok(());
+    }
+
+    let mode = match module_hash {
+        None => Install,
+        Some(_) => Upgrade,
+    };
+
+    let settings = ArchiveInit {
+        ii_canister: id(),
+        max_entries_per_call: 1000,
+    };
+    let encoded_arg = candid::encode_one(settings)
+        .map_err(|err| format!("failed to encode archive install argument: {:?}", err))?;
+
+    install_code(InstallCodeArgument {
+        mode,
+        canister_id: archive_canister,
+        wasm_module,
+        arg: encoded_arg,
+    })
+    .await
+    .map_err(|err| format!("failed to install archive canister: {:?}", err))
+}
+
+pub fn archive_operation(anchor: UserNumber, caller: Principal, operation: Operation) {
+    let archive_data = state::archive_data(); // traps if archive is not available
+    let timestamp = time();
+    let entry = Entry {
+        anchor,
+        operation,
+        timestamp,
+        caller,
+        sequence_number: archive_data.sequence_number,
+    };
+    let encoded_entry = candid::encode_one(entry).expect("failed to encode archive entry");
```

Comment:
out of curiosity: I'm really surprised that _encoding_ may fail. Do you know why that might be the case?

Reply:
As explained by @roman-kashitsyn:
> All serde-based encoders can fail. One reason is that serde tries to serialize values under a mutex, and grabbing a mutex might fail. Not a great design decision IMO; serialization should be infallible.